{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set the environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "from cobra import Model, Reaction, Metabolite\n",
    "from cobra.flux_analysis import flux_variability_analysis\n",
    "import math\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import glob\n",
    "import escher\n",
    "from escher import Builder\n",
    "from utils import show_map\n",
    "from utils.check_precursor_problem import check_precursor_problem"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read the excel that have all gene reaction rule information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene_reaction_rule =pd.read_excel('../data/gene reaction rule.xlsx')\n",
    "# set the reaction id column as the index in order to map it to the model \n",
    "df_gene_reaction_rule = df_gene_reaction_rule.set_index([\"reaction_id\"],)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read the raw transcriptomics of temperature treatment  70 and 80 ℃"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "xl = pd.ExcelFile('../data/80vs70 ArrayData.xlsx') # read an excel file\n",
    "names1 = xl.sheet_names  # write all the sheet names in one list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Format all the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>5</th>\n",
       "      <th>R</th>\n",
       "      <th>C</th>\n",
       "      <th>Name</th>\n",
       "      <th>ID</th>\n",
       "      <th>Array_67 70</th>\n",
       "      <th>Array_67 80</th>\n",
       "      <th>Array_68 70</th>\n",
       "      <th>Array_68 80</th>\n",
       "      <th>Array_86 70</th>\n",
       "      <th>Array_86 80</th>\n",
       "      <th>Array_85 70</th>\n",
       "      <th>Array_85 80</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SSO number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gi|16206:1-933_A.thaliana_Cab_mRNA_oligo1_553-608</td>\n",
       "      <td>116020</td>\n",
       "      <td>85320</td>\n",
       "      <td>118800</td>\n",
       "      <td>124647</td>\n",
       "      <td>86190</td>\n",
       "      <td>145044</td>\n",
       "      <td>163856</td>\n",
       "      <td>89737</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gi|16206:1-933_A.thaliana_Cab_mRNA_oligo1_553-608</td>\n",
       "      <td>119282</td>\n",
       "      <td>96720</td>\n",
       "      <td>129168</td>\n",
       "      <td>127391</td>\n",
       "      <td>99104</td>\n",
       "      <td>157584</td>\n",
       "      <td>119132</td>\n",
       "      <td>87409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gi|16470|emb|X14212.1|_A.thaliana_RCA_mRNA_oligo2_1232-1288</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>65521</td>\n",
       "      <td>149152</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gi|16470|emb|X14212.1|_A.thaliana_RCA_mRNA_oligo2_1232-1288</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gi|1928871|gb|U91966.1|ATU91966_A.thaliana_rbcL_mRNA_oligo3_1088-1155</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>69295</td>\n",
       "      <td>155472</td>\n",
       "      <td>162288</td>\n",
       "      <td>59350</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "5           R  C Name                                                                     ID Array_67 70 Array_67 80 Array_68 70 Array_68 80 Array_86 70 Array_86 80 Array_85 70 Array_85 80\n",
       "SSO number                                                                                                                                                                                  \n",
       "            1  1  NaN  gi|16206:1-933_A.thaliana_Cab_mRNA_oligo1_553-608                      116020      85320       118800      124647      86190       145044      163856      89737     \n",
       "            1  2  NaN  gi|16206:1-933_A.thaliana_Cab_mRNA_oligo1_553-608                      119282      96720       129168      127391      99104       157584      119132      87409     \n",
       "            1  3  NaN  gi|16470|emb|X14212.1|_A.thaliana_RCA_mRNA_oligo2_1232-1288                                                            65521       149152                            \n",
       "            1  4  NaN  gi|16470|emb|X14212.1|_A.thaliana_RCA_mRNA_oligo2_1232-1288                                                                                                          \n",
       "            1  5  NaN  gi|1928871|gb|U91966.1|ATU91966_A.thaliana_rbcL_mRNA_oligo3_1088-1155                                                  69295       155472      162288      59350     "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_overall_data (names1):\n",
    "    df_overall_data = xl.parse(\"Array_Annotation\") # write sheet named \"Array_Annotation\" in excel into a dataframe\n",
    "    df_overall_data = df_overall_data.drop(labels=[0,1,2,3,4],axis = 0) # drop the first 5 lines\n",
    "    df_overall_data.columns = df_overall_data.iloc[0] # set the first line as the column names\n",
    "    df_overall_data = df_overall_data.drop(labels=[5],axis = 0) # delete the line with label 5\n",
    "    df_overall_data = df_overall_data.set_index([\"UID\"],) # set the index\n",
    "    df_overall_data[\"SSO number\"] = \"\" # add a new blank column with name 'SSO number'\n",
    "    for  UID in df_overall_data.index:\n",
    "        if \"SSO\" in df_overall_data[\"ID\"].loc[UID]:\n",
    "            df_overall_data[\"SSO number\"].at[UID] = str(df_overall_data[\"ID\"].loc[UID]).split(\"_\")[0]+\"_\" # the reason for adding \"_\" is because some SSO numbers \n",
    "            #only has 4 numbers and others has 5 numbers, to avoide the conditions like this (2133 and 21338)\n",
    "    for name in names1:\n",
    "        if name not in [\"Array_Annotation\", \"Metadata\"]:# the sheet name in the list cannot be processed by the following codes\n",
    "            df = xl.parse(name)\n",
    "            df  = df.drop(labels=[0,1,2,3,4,5],axis = 0)\n",
    "            df.columns = df.iloc[0]\n",
    "            df = df.drop(labels=[6],axis = 0)\n",
    "            df = df.set_index([\"UID\"],)\n",
    "            column_name_70 = name + \" 70\" #set the column name to sheet name add temperature\n",
    "            column_name_80 = name + \" 80\"\n",
    "            df_overall_data[column_name_70] = \"\"\n",
    "            df_overall_data[column_name_80] = \"\" # set empty column\n",
    "\n",
    "            for i in set(df_overall_data.index).intersection(set(df.index)): # find the same items in two set\n",
    "\n",
    "                df_overall_data[column_name_70].at[i] = df[\"70°C\"].loc[i]\n",
    "                df_overall_data[column_name_80].at[i] = df[\"80°C\"].loc[i]\n",
    "        \n",
    "    return (df_overall_data)\n",
    "df_overall_data = get_overall_data(names1)\n",
    "df_overall_data = df_overall_data.set_index([\"SSO number\"],)\n",
    "df_overall_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Match transcriptome to the reaction id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\yzhang2\\Anaconda3\\envs\\archeapy36\\lib\\site-packages\\pandas\\core\\indexing.py:205 SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reaction name</th>\n",
       "      <th>pathway</th>\n",
       "      <th>genes</th>\n",
       "      <th>gene_reaction_rule</th>\n",
       "      <th>Unnamed: 5</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "      <th>Array_67 70</th>\n",
       "      <th>Array_67 80</th>\n",
       "      <th>Array_68 70</th>\n",
       "      <th>Array_68 80</th>\n",
       "      <th>Array_86 70</th>\n",
       "      <th>Array_86 80</th>\n",
       "      <th>Array_85 70</th>\n",
       "      <th>Array_85 80</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reaction_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>T_ABC__transporter_beta__D__glucose</td>\n",
       "      <td>ABC transport</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2847,3066,2848,3067,2849,3068,2850,3069</td>\n",
       "      <td>and</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>142229</td>\n",
       "      <td>63180</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>196101</td>\n",
       "      <td>99840</td>\n",
       "      <td>215040</td>\n",
       "      <td>141324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>carb_entner_1.1.1.359_GLUCOSE__1__DEHYDROGENASE__NAD+__RXN</td>\n",
       "      <td>GDH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3003,3042,3204</td>\n",
       "      <td>or</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>105265</td>\n",
       "      <td>111072</td>\n",
       "      <td>134460</td>\n",
       "      <td>107596</td>\n",
       "      <td>134495</td>\n",
       "      <td>154112</td>\n",
       "      <td>150784</td>\n",
       "      <td>92612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>carb_entner_1.1.1.47_GLUCOSE__1__DEHYDROGENASE__NAD+__RXN</td>\n",
       "      <td>GDH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3003,3042,3204</td>\n",
       "      <td>or</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>105265</td>\n",
       "      <td>111072</td>\n",
       "      <td>134460</td>\n",
       "      <td>107596</td>\n",
       "      <td>134495</td>\n",
       "      <td>154112</td>\n",
       "      <td>150784</td>\n",
       "      <td>92612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>carb_entner_1.1.1.360_GLUCOSE__1__DEHYDROGENASE__NADP+__RXN</td>\n",
       "      <td>GDH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3003,3042,3204</td>\n",
       "      <td>or</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>105265</td>\n",
       "      <td>111072</td>\n",
       "      <td>134460</td>\n",
       "      <td>107596</td>\n",
       "      <td>134495</td>\n",
       "      <td>154112</td>\n",
       "      <td>150784</td>\n",
       "      <td>92612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>carb_entner_3.1.1.17_GLUCONOLACT__RXN</td>\n",
       "      <td>GL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>spontaneous</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1000000</td>\n",
       "      <td>1000000</td>\n",
       "      <td>1000000</td>\n",
       "      <td>1000000</td>\n",
       "      <td>1000000</td>\n",
       "      <td>1000000</td>\n",
       "      <td>1000000</td>\n",
       "      <td>1000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             Reaction name pathway                                    genes gene_reaction_rule  Unnamed: 5 Unnamed: 6 Array_67 70 Array_67 80 Array_68 70 Array_68 80 Array_86 70 Array_86 80 Array_85 70 Array_85 80\n",
       "reaction_id                                                                                                                                                                                                                                                          \n",
       "T_ABC__transporter_beta__D__glucose                          ABC transport  NaN     2847,3066,2848,3067,2849,3068,2850,3069  and               NaN          NaN        142229      63180                               196101      99840       215040      141324    \n",
       "carb_entner_1.1.1.359_GLUCOSE__1__DEHYDROGENASE__NAD+__RXN   GDH            NaN     3003,3042,3204                           or                NaN          NaN        105265      111072      134460      107596      134495      154112      150784      92612     \n",
       "carb_entner_1.1.1.47_GLUCOSE__1__DEHYDROGENASE__NAD+__RXN    GDH            NaN     3003,3042,3204                           or                NaN          NaN        105265      111072      134460      107596      134495      154112      150784      92612     \n",
       "carb_entner_1.1.1.360_GLUCOSE__1__DEHYDROGENASE__NADP+__RXN  GDH            NaN     3003,3042,3204                           or                NaN          NaN        105265      111072      134460      107596      134495      154112      150784      92612     \n",
       "carb_entner_3.1.1.17_GLUCONOLACT__RXN                        GL             NaN     spontaneous                              NaN               NaN          NaN        1000000     1000000     1000000     1000000     1000000     1000000     1000000     1000000   "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def match_data_to_reaction_id(columns): \n",
    "    \n",
    "    for column_name in columns:\n",
    "        if \"Array\" in column_name: # select the column with \"Array\" in its name\n",
    "            \n",
    "            df_gene_reaction_rule[column_name] = \"\"\n",
    "            # the following list is special because its items' reaction rule is special with both \"and\" and \"or\" in it\n",
    "            list_id_with_special_gene_reaction_rule = [\"carb_acetylcoa_1.2.7.1_PYRUFLAVREDUCT__RXN\",\"oxp_redox_1.6.5.3_RXN0__5330\",'carb_degra__gly_1.8.1.4_1.4.4.2_2.1.2.10_GCVMULTI__RXN']\n",
    "            for rxn in df_gene_reaction_rule.index:\n",
    "                \n",
    "                if rxn not in list_id_with_special_gene_reaction_rule:\n",
    "                    if 'spontaneous' not in str(df_gene_reaction_rule['genes'].loc[rxn]):\n",
    "                        c = str(df_gene_reaction_rule[\"genes\"].loc[rxn]).split(\",\")\n",
    "                       \n",
    "                        if len(c) > 1:# select the reactions catalyzed by enzymes with more subunits or isoenzymes\n",
    "\n",
    "                            gene_numbers1 = []\n",
    "                            \n",
    "                            if \"or\" in str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]):\n",
    "\n",
    "                                for i in range(len(c)):\n",
    "                                    e = \"SSO\" + c[i]+\"_\"\n",
    "\n",
    "                                    if e in df_overall_data.index:\n",
    "                                        \n",
    "                                        if df_overall_data[column_name].loc[e] !='':\n",
    "                                            gene_numbers1.append(df_overall_data[column_name].loc[e])\n",
    "     \n",
    "\n",
    "                            df_gene_reaction_rule[column_name].loc[rxn] = sum(gene_numbers1)\n",
    "                    \n",
    "\n",
    "                            if \"and\" in str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]):               \n",
    "\n",
    "                                for i in range(len(c)):\n",
    "                                    e = \"SSO\" + c[i]+\"_\"\n",
    "\n",
    "                                    if e in df_overall_data.index:\n",
    "\n",
    "                                        gene_numbers1.append(df_overall_data[column_name].loc[e])\n",
    "\n",
    "                                       \n",
    "                            if '' in gene_numbers1:\n",
    "                                df_gene_reaction_rule[column_name].loc[rxn] = \"\"\n",
    "                             \n",
    "                            else:\n",
    "                                df_gene_reaction_rule[column_name].loc[rxn] = min(gene_numbers1)\n",
    "                             \n",
    "\n",
    "                        else:\n",
    "                            for i in range(len(c)):\n",
    "                                e = \"SSO\" + c[i]+\"_\"\n",
    "\n",
    "                                if e in df_overall_data.index:\n",
    "                                    df_gene_reaction_rule[column_name].loc[rxn] = df_overall_data[column_name].loc[e]  \n",
    "                                    df_gene_reaction_rule[column_name].loc[rxn] = df_overall_data[column_name].loc[e]\n",
    "                    else:\n",
    "                        df_gene_reaction_rule[column_name].loc[rxn] = 1000000 # if spontaneous in it, set the number to 1000000.\n",
    "                        df_gene_reaction_rule[column_name].loc[rxn] = 1000000\n",
    "                elif rxn in [\"oxp_redox_1.6.5.3_RXN0__5330\",\"carb_acetylcoa_1.2.7.1_PYRUFLAVREDUCT__RXN\"]:\n",
    "                    g = df_gene_reaction_rule[\"genes\"].loc[rxn].split(\"or\")\n",
    "                   \n",
    "                    gene_numbers1 = []\n",
    "                    \n",
    "                    c1 = str(g[0]).split(\",\")\n",
    "                    for i in range(len(c1)):\n",
    "\n",
    "                        e = \"SSO\" + c1[i]+ \"_\"\n",
    "\n",
    "                        if e in df_overall_data.index:\n",
    "                            \n",
    "                            gene_numbers1.append(df_overall_data[column_name].loc[e])\n",
    "                                \n",
    "\n",
    "\n",
    "                    gene_numbers3 = []\n",
    "                   \n",
    "                    c2 = str(g[1]).split(\",\")\n",
    "                    for i in range(len(c2)):\n",
    "\n",
    "                        e = \"SSO\" + c2[i]+\"_\"\n",
    "                        if e in df_overall_data.index:\n",
    "                        \n",
    "                            gene_numbers3.append(df_overall_data[column_name].loc[e])\n",
    "\n",
    "\n",
    "                    if '' not in gene_numbers3:\n",
    "                        if '' in gene_numbers1:\n",
    "                            df_gene_reaction_rule[column_name].loc[rxn] = min(gene_numbers3)\n",
    "                        else:\n",
    "                            df_gene_reaction_rule[column_name].loc[rxn] = min(gene_numbers3) + min(gene_numbers3)\n",
    "                    else:\n",
    "                        \n",
    "                        if '' in gene_numbers1:\n",
    "                            df_gene_reaction_rule[column_name].loc[rxn] = ''\n",
    "                        else:\n",
    "                            df_gene_reaction_rule[column_name].loc[rxn] = min(gene_numbers1)\n",
    "                    \n",
    "                    \n",
    "                elif rxn == \"carb_degra__gly_1.8.1.4_1.4.4.2_2.1.2.10_GCVMULTI__RXN\":\n",
    "                    g = str(df_gene_reaction_rule[\"genes\"].loc[rxn]).split(\")\")\n",
    "                    \n",
    "                    gene_numbers1 = []\n",
    "                    \n",
    "                    c1 = str(g[0]).split(\",\")\n",
    "                    c1[0] = c1[0][1:]\n",
    "\n",
    "                    for i in range(len(c1)):\n",
    "\n",
    "                        e = \"SSO\" + c1[i]+ \"_\"\n",
    "\n",
    "                        if e in df_overall_data.index:\n",
    "\n",
    "                            if df_overall_data[column_name].loc[e] !='':\n",
    "                                gene_numbers1.append(df_overall_data[column_name].loc[e])\n",
    "                            \n",
    "    \n",
    "                    gene_numbers3 = []\n",
    "                    \n",
    "                    c2 = str(g[1]).split(\",\")\n",
    "                    c2[0] = c2[0][1:]\n",
    "                    for i in range(len(c2)):\n",
    "\n",
    "                        e = \"SSO\" + c2[i]+\"_\"\n",
    "                        if e in df_overall_data.index:\n",
    "\n",
    "                            gene_numbers3.append(df_overall_data[column_name].loc[e])\n",
    "                          \n",
    "\n",
    "                    \n",
    "                    \n",
    "                    if '' in gene_numbers3:\n",
    "                         df_gene_reaction_rule.loc[rxn,column_name] = ''\n",
    "                    else:\n",
    "                        score = min(gene_numbers3)\n",
    "                        df_gene_reaction_rule.loc[rxn,column_name] = min(sum(gene_numbers1),score)\n",
    "\n",
    "    return (df_gene_reaction_rule)\n",
    "list_columns = df_overall_data.columns\n",
    "df_match_data_to_reaction_id = match_data_to_reaction_id(list_columns)\n",
    "df_match_data_to_reaction_id.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate the average"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_match_data_to_reaction_id.to_csv(\"../data/matched_transcriptone_data_to_reaction_id.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_match_data_to_reaction_id =pd.read_csv('../data/matched_transcriptone_data_to_reaction_id.csv')\n",
    "list_columns_70 = [i for i in df_match_data_to_reaction_id.columns if \" 70\" in i]\n",
    "list_columns_80 = [i for i in df_match_data_to_reaction_id.columns if \" 80\" in i]\n",
    "df_match_data_to_reaction_id[\"average_list_column_70\"] = df_match_data_to_reaction_id[list_columns_70].mean(axis= 1) # calculate the average values of different columns\n",
    "df_match_data_to_reaction_id[\"average_list_column_80\"] = df_match_data_to_reaction_id[list_columns_80].mean(axis= 1)\n",
    "df_match_data_to_reaction_id.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_match_data_to_reaction_id.to_csv(\"../data/reaction_id_RAS.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model1= model.copy()\n",
    "df = pd.DataFrame()\n",
    "df[\"soxM\"] = \"\"\n",
    "df[\"soxABCD\"] = \"\"\n",
    "df[\"doxABCDE\"] = \"\"\n",
    "df[\"soxM percent\"] = \"\"\n",
    "df[\"soxABCD percent\"] = \"\"\n",
    "df[\"doxABCDE percent\"] = \"\"\n",
    "for a in [1,3,5,7,9,11,13,15,100]:\n",
    "    for rxn in df_gene_reaction_rule.index: \n",
    "        if \".\" in str(df_gene_reaction_rule['Array_67_70°C'].loc[rxn]):\n",
    "\n",
    "            the_bound = 0.001*df_gene_reaction_rule['Array_67_70°C'].loc[rxn]/a\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, the_bound)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, the_bound)  # the lower bound will stay zero\n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, 0) # the upper bound will stay zero\n",
    "        else:\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 1000)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            if model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, 1000)  # the lower bound will stay zero\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 0) \n",
    "    model1.objective  = model1.reactions.NGAM\n",
    "    sol = model1.optimize()\n",
    "    df.loc[a] = [sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'])]    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HU treatment  transcript"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### excel that contains the HU treatment  transcript\n",
    "- The data is from the paper 'HU treatment of Sulfolobus solfataricus P2'.\n",
    "- Samples were taken from the cells growing with or without the treatment of Hydroxyurea (HU) for 4 hours.\n",
    "- The data can be download from the website (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE87612)\n",
    "- As this transcript data doesn't contain the information of SSO numbers, and it contains the gene location information so we used a file with SSO numbers and also gene location information to map the SSO numbers to the transcript data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add tge gene reaction rule to the model\n",
    "# reactions_ids = [rxn.id for rxn in model.reactions]\n",
    "df_genome=pd.read_table('../data/GCA_000007005.1_ASM700v1_feature_table.txt') # This file contains the gene information of sulfolubus. \n",
    "df_HU_treatment =pd.read_excel('../data/transciptone.xlsx',sheet_name = \"HU treatment of sulfolobus \")\n",
    "df_genome = df_genome.set_index([\"# feature\"],)\n",
    "df_genome = df_genome.drop(index='gene')\n",
    "df_HU_treatment = df_HU_treatment.set_index([\"START\"],)\n",
    "df_HU_treatment[\"ORF\"] = \"\"\n",
    "for start in df_HU_treatment.index:\n",
    "    for i in range(len(df_genome)):\n",
    "        if start == df_genome[\"start\"].iloc[i]: \n",
    "            df_HU_treatment[\"ORF\"].at[start] = (df_genome[\"locus_tag\"].iloc[i]).strip()     \n",
    "df_HU_treatment = df_HU_treatment.set_index(\"ORF\")\n",
    "df_gene_reaction_rule[\"HU TREATED/CONTROL\"] = \"\"\n",
    "df_gene_reaction_rule[\"RAS_HU TREATED/CONTROL\"] = \"\"\n",
    "\n",
    "df_gene_reaction_rule[\"gene_reaction_relation\"] = \"\"\n",
    "for rxn in df_gene_reaction_rule.index:\n",
    "    \n",
    "    if rxn != \"oxp_redox_1.6.5.3_RXN0__5330\":\n",
    "        if 'spontaneous' not in str(df_gene_reaction_rule['genes'].loc[rxn]):\n",
    "            c = df_gene_reaction_rule[\"genes\"].loc[rxn].split(\",\")\n",
    "            if len(c) > 1:\n",
    "                gene_rules = []\n",
    "                gene_numbers1 = []\n",
    "                gene_numbers2 = []\n",
    "                \n",
    "                for i in range(len(c)):\n",
    "                    e = \"SSO\" + c[i]\n",
    "                    if e in df_HU_treatment.index:\n",
    "#                         print (e)\n",
    "                        gene_rules.append(e)\n",
    "#                         ge = df_HU_treatment[\"HU TREATED\"].loc[e]/df_HU_treatment[\"CONTROL\"].loc[e]\n",
    "                       \n",
    "                        gene_numbers1.append(df_HU_treatment[\"HU TREATED/CONTROL\"].loc[e])\n",
    "#                         gene_numbers2.append(df_HU_treatment[\"CONTROL\"].loc[e])\n",
    "                  \n",
    "                if \"or\" in df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]:\n",
    "                    print (gene_numbers1)\n",
    "                    df_gene_reaction_rule[\"RAS_HU TREATED/CONTROL\"].loc[rxn] = sum(gene_numbers1)\n",
    "\n",
    "#                     df_gene_reaction_rule[\"RAS_control\"].loc[rxn] = sum(gene_numbers2)\n",
    "#                 if \"and\" in str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]):\n",
    "#                     df_gene_reaction_rule[\"RAS_HU TREATED/CONTROL\"].loc[rxn] = min(gene_numbers1)\n",
    "\n",
    "#                     df_gene_reaction_rule[\"RAS_control\"].loc[rxn] = min(gene_numbers2)\n",
    "                    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add tge gene reaction rule to the model\n",
    "# reactions_ids = [rxn.id for rxn in model.reactions]\n",
    "df_genome=pd.read_table('../data/GCA_000007005.1_ASM700v1_feature_table.txt') # This file contains the gene information of sulfolubus. \n",
    "df_HU_treatment =pd.read_excel('../data/transciptone.xlsx',sheet_name = \"HU treatment of sulfolobus \")\n",
    "df_genome = df_genome.set_index([\"# feature\"],)\n",
    "df_genome = df_genome.drop(index='gene')\n",
    "df_HU_treatment = df_HU_treatment.set_index([\"START\"],)\n",
    "df_HU_treatment[\"ORF\"] = \"\"\n",
    "for start in df_HU_treatment.index:\n",
    "    for i in range(len(df_genome)):\n",
    "        if start == df_genome[\"start\"].iloc[i]: \n",
    "            df_HU_treatment[\"ORF\"].at[start] = (df_genome[\"locus_tag\"].iloc[i]).strip()     \n",
    "df_HU_treatment = df_HU_treatment.set_index(\"ORF\")\n",
    "df_gene_reaction_rule[\"HU TREATED/CONTROL\"] = \"\"\n",
    "df_gene_reaction_rule[\"RAS_HU TREATED/CONTROL\"] = \"\"\n",
    "\n",
    "df_gene_reaction_rule[\"gene_reaction_relation\"] = \"\"\n",
    "for rxn in df_gene_reaction_rule.index:\n",
    "    \n",
    "    if rxn != \"oxp_redox_1.6.5.3_RXN0__5330\":\n",
    "        if 'spontaneous' not in str(df_gene_reaction_rule['genes'].loc[rxn]):\n",
    "            c = df_gene_reaction_rule[\"genes\"].loc[rxn].split(\",\")\n",
    "            if len(c) > 1:\n",
    "                gene_rules = []\n",
    "                gene_numbers1 = []\n",
    "                gene_numbers2 = []\n",
    "                \n",
    "                for i in range(len(c)):\n",
    "                    e = \"SSO\" + c[i]\n",
    "                    if e in df_HU_treatment.index:\n",
    "#                         print (e)\n",
    "                        gene_rules.append(e)\n",
    "#                         ge = df_HU_treatment[\"HU TREATED\"].loc[e]/df_HU_treatment[\"CONTROL\"].loc[e]\n",
    "                       \n",
    "                        gene_numbers1.append(df_HU_treatment[\"HU TREATED/CONTROL\"].loc[e])\n",
    "#                         gene_numbers2.append(df_HU_treatment[\"CONTROL\"].loc[e])\n",
    "                  \n",
    "                if \"or\" in df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]:\n",
    "                    df_gene_reaction_rule[\"RAS_HU TREATED/CONTROL\"].loc[rxn] = sum(gene_numbers1)\n",
    "\n",
    "#                     df_gene_reaction_rule[\"RAS_control\"].loc[rxn] = sum(gene_numbers2)\n",
    "                if \"and\" in str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]):\n",
    "                    df_gene_reaction_rule[\"RAS_HU TREATED/CONTROL\"].loc[rxn] = min(gene_numbers1)\n",
    "\n",
    "#                     df_gene_reaction_rule[\"RAS_control\"].loc[rxn] = min(gene_numbers2)\n",
    "                    \n",
    "\n",
    "        \n",
    "                d = str(\" \"+ str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]) + \" \").join(gene_rules)\n",
    "                df_gene_reaction_rule[\"HU TREATED/CONTROL\"].loc[rxn] = gene_numbers1\n",
    "#                 df_gene_reaction_rule[\"control\"].loc[rxn] = gene_numbers2\n",
    "                df_gene_reaction_rule[\"gene_reaction_relation\"].loc[rxn] = \"(\" + d + ')'\n",
    "            elif len(c) == 1:\n",
    "                gene_rules = []\n",
    "                gene_numbers1 = []\n",
    "                gene_numbers2 = []\n",
    "                \n",
    "                e = \"SSO\" + c[0]\n",
    "                if e in df_HU_treatment.index:\n",
    "                    d = e \n",
    "                    gene_numbers1.append(df_HU_treatment[\"HU TREATED/CONTROL\"].loc[e])\n",
    "#                     gene_numbers2.append(df_HU_treatment[\"CONTROL\"].loc[e])\n",
    "                   \n",
    "                    df_gene_reaction_rule[\"HU TREATED/CONTROL\"].loc[rxn] = gene_numbers1\n",
    "#                     df_gene_reaction_rule[\"control\"].loc[rxn] = gene_numbers2\n",
    "                    df_gene_reaction_rule[\"gene_reaction_relation\"].loc[rxn] = d\n",
    "                    df_gene_reaction_rule[\"RAS_HU TREATED/CONTROL\"].loc[rxn] = df_HU_treatment[\"HU TREATED/CONTROL\"].loc[e]\n",
    "#                     df_gene_reaction_rule[\"RAS_control\"].loc[rxn] = df_HU_treatment[\"CONTROL\"].loc[e]\n",
    "        else:\n",
    "            \n",
    "            df_gene_reaction_rule[\"HU TREATED/CONTROL\"].loc[rxn] = 1000\n",
    "#             df_gene_reaction_rule[\"RAS_control\"].loc[rxn] = 1000\n",
    "                    \n",
    "           \n",
    "    elif rxn == \"oxp_redox_1.6.5.3_RXN0__5330\":\n",
    "        g = df_gene_reaction_rule[\"genes\"].loc[rxn].split(\"or\")\n",
    "        gene_rules = []\n",
    "        gene_numbers1 = []\n",
    "        gene_numbers2 = []\n",
    "        c1 = g[0].split(\",\")\n",
    "        for i in range(len(c1)):\n",
    "        \n",
    "            e = \"SSO\" + c1[i]\n",
    "            if e in df_HU_treatment.index:\n",
    "#  \n",
    "                gene_rules.append(e)\n",
    "                gene_numbers1.append(df_HU_treatment[\"HU TREATED/CONTROL\"].loc[e])\n",
    "#                 gene_numbers2.append(df_HU_treatment[\"CONTROL\"].loc[e])\n",
    "        \n",
    "               \n",
    "        \n",
    "               \n",
    "            \n",
    "        d1 = str(\" \"+ str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]) + \" \").join(gene_rules)\n",
    "        gene_numbers3 = []\n",
    "        gene_numbers4 = []\n",
    "        c2 = g[1].split(\",\")\n",
    "        for i in range(len(c2)):\n",
    "        \n",
    "            e = \"SSO\" + c2[i]\n",
    "            if e in df_HU_treatment.index:  \n",
    "                gene_numbers3.append(df_HU_treatment[\"HU TREATED/CONTROL\"].loc[e])\n",
    "#                 gene_numbers4.append(df_HU_treatment[\"CONTROL\"].loc[e])\n",
    "        \n",
    "               \n",
    "        df_gene_reaction_rule[\"HU TREATED/CONTROL\"].loc[rxn] = gene_numbers1 + gene_numbers3\n",
    "#         df_gene_reaction_rule[\"control\"].loc[rxn] = gene_numbers2 + gene_numbers4\n",
    "        d2 = str(\" \"+ str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]) + \" \").join(gene_rules)\n",
    "        \n",
    "        df_gene_reaction_rule[\"gene_reaction_relation\"].loc[rxn] = \"((\" + d1 + \") or (\" +  d2+ \"))\" \n",
    "        df_gene_reaction_rule[\"RAS_HU TREATED/CONTROL\"].loc[rxn] = min(gene_numbers1) + min(gene_numbers3)\n",
    "#         df_gene_reaction_rule[\"RAS_control\"].loc[rxn] = min(gene_numbers2) + min(gene_numbers4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene_reaction_rule"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene_reaction_rule [\"RAS_HU TREATED/CONTROL\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RAS_HU TREATED/CONTROL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model1 = M.copy()\n",
    "df = pd.DataFrame()\n",
    "df[\"soxM\"] = \"\"\n",
    "df[\"soxABCD\"] = \"\"\n",
    "df[\"doxABCDE\"] = \"\"\n",
    "df[\"soxM percent\"] = \"\"\n",
    "df[\"soxABCD percent\"] = \"\"\n",
    "df[\"doxABCDE percent\"] = \"\"\n",
    "df[\"soxM bounds\"] = \"\"\n",
    "df[\"soxABCD bounds\"] = \"\"\n",
    "df[\"doxABCDE bounds\"] = \"\"\n",
    "df[\"ATPase bounds\"] = \"\"\n",
    "df[\"NGAM\"] = \"\"\n",
    "df[\"ENO\"] = \"\"\n",
    "df[\"ATPase\"] = \"\"\n",
    "for a in [1,3,5,7,9,11,13,15]:\n",
    "    for rxn in df_gene_reaction_rule.index: \n",
    "        if \".\" in str(df_gene_reaction_rule['RAS_HU TREATED/CONTROL'].loc[rxn]):\n",
    "\n",
    "            the_bound = a*df_gene_reaction_rule['RAS_HU TREATED/CONTROL'].loc[rxn]\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, the_bound)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, the_bound)  # the lower bound will stay zero\n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, 0) # the upper bound will stay zero\n",
    "        else:\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 1000)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            if model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, 1000)  # the lower bound will stay zero\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 0) \n",
    "    model1.objective  = model1.reactions.NGAM\n",
    "    sol = model1.optimize()\n",
    "#     df.loc[a] = [sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),model1.reactions.get_by_id('oxp_redox_1.10.3.13_soxM_TRANS__RXN__240').bounds,model1.reactions.get_by_id('oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240').bounds,model1.reactions.get_by_id('oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240').bounds,sol.fluxes.loc['NGAM'],sol.fluxes.loc['carb_enter_glucneo_4.2.1.11_2PGADEHYDRAT__RXN'],sol.fluxes.loc['oxp_3.6.3.14_ATP__synthase']]        \n",
    "    df.loc[a] = [sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),model1.reactions.get_by_id('oxp_redox_1.10.3.13_soxM_TRANS__RXN__240').bounds,model1.reactions.get_by_id('oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240').bounds,model1.reactions.get_by_id('oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240').bounds,model1.reactions.get_by_id('oxp_3.6.3.14_ATP__synthase').bounds,sol.fluxes.loc['NGAM'],sol.fluxes.loc['carb_enter_glucneo_4.2.1.11_2PGADEHYDRAT__RXN'],sol.fluxes.loc['oxp_3.6.3.14_ATP__synthase']]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for rxn in df_gene_reaction_rule.index: \n",
    "    if \".\" in str(df_gene_reaction_rule['RAS_HU TREATED/CONTROL'].loc[rxn]):\n",
    "\n",
    "        the_bound = 9*df_gene_reaction_rule['RAS_HU TREATED/CONTROL'].loc[rxn]\n",
    "\n",
    "        if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, the_bound)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "        elif model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (0, the_bound)  # the lower bound will stay zero\n",
    "\n",
    "        elif model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, 0) # the upper bound will stay zero\n",
    "    else:\n",
    "        if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (-1000, 1000)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "        if model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (0, 1000)  # the lower bound will stay zero\n",
    "\n",
    "        if model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (-1000, 0) \n",
    "model1.objective  = model1.reactions.NGAM\n",
    "sol = model1.optimize()\n",
    "l = sol.fluxes[abs(sol.fluxes)>1e-6].index\n",
    "\n",
    "for rxn in l:\n",
    "    print (rxn,model1.reactions.get_by_id(rxn).reaction,model1.reactions.get_by_id(rxn).bounds,sol.fluxes.loc[rxn])\n",
    "b = show_map(sol,map_loc)\n",
    "b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model1.reactions.T_flux_ammonium.bounds"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Temperature_treatment transcript \n",
    "- Import the excel that contain the transcript data with shifting temperature from 80 ℃ to 90 ℃.\n",
    "- The data is from the paper 'Dynamic Metabolic Adjustments and Genome Plasticity Are Implicated in the Heat Shock Response of the Extremely Thermoacidophilic Archaeon Sulfolobus solfataricus'.\n",
    "- Samples were taken 10 min (-10) before starting the temperature shift and then 5, 30, and 60 min after reaching 90°C.\n",
    "- The data can be download from the website (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1482968/bin/jbacter_188_12_4553__index.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene_reaction_rule[\"Array_67_70°C\"] = \"\"\n",
    "df_gene_reaction_rule[\"Array_67_80°C\"] = \"\"\n",
    "\n",
    "df_gene_reaction_rule[\"gene_reaction_relation\"] = \"\"\n",
    "for rxn in df_gene_reaction_rule.index:\n",
    "    \n",
    "    if rxn != \"oxp_redox_1.6.5.3_RXN0__5330\":\n",
    "        if 'spontaneous' not in str(df_gene_reaction_rule['genes'].loc[rxn]):\n",
    "            c = df_gene_reaction_rule[\"genes\"].loc[rxn].split(\",\")\n",
    "            if len(c) > 1:\n",
    "                gene_rules = []\n",
    "                gene_numbers1 = []\n",
    "                gene_numbers2 = []\n",
    "                \n",
    "                for i in range(len(c)):\n",
    "                    e = \"SSO\" + c[i]\n",
    "                    if e in df_temperature_treatment.index:\n",
    "                        print (e)\n",
    "                        gene_numbers1.append(df_temperature_treatment[\"Array_67_70°C\"].loc[e])\n",
    "                        gene_numbers2.append(df_temperature_treatment[\"Array_67_80°C\"].loc[e])\n",
    "                  \n",
    "                if \"or\" in df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]:\n",
    "                    df_gene_reaction_rule[\"Array_67_70°C\"].loc[rxn] = sum(gene_numbers1)\n",
    "                    df_gene_reaction_rule[\"Array_67_80°C\"].loc[rxn] = sum(gene_numbers2)\n",
    "                if \"and\" in str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]):\n",
    "                    df_gene_reaction_rule[\"Array_67_70°C\"].loc[rxn] = min(gene_numbers1)\n",
    "                    df_gene_reaction_rule[\"Array_67_80°C\"].loc[rxn] = min(gene_numbers2)\n",
    "                    \n",
    "\n",
    "            elif len(c) == 1:\n",
    "                gene_rules = []\n",
    "                gene_numbers1 = []\n",
    "                gene_numbers2 = []\n",
    "                \n",
    "                e = \"SSO\" + c[0]\n",
    "                if e in df_temperature_treatment.index:\n",
    "                    d = e \n",
    "                    gene_numbers1.append(df_temperature_treatment[\"Array_67_70°C\"].loc[e])\n",
    "                    gene_numbers2.append(df_temperature_treatment[\"Array_67_80°C\"].loc[e])\n",
    "                    df_gene_reaction_rule[\"Array_67_70°C\"].loc[rxn] = df_temperature_treatment[\"Array_67_80°C\"].loc[e]\n",
    "                    df_gene_reaction_rule[\"Array_67_80°C\"].loc[rxn] = df_temperature_treatment[\"Array_67_80°C\"].loc[e]\n",
    "        else:\n",
    "            \n",
    "            df_gene_reaction_rule[\"Array_67_70°C\"].loc[rxn] = 1000\n",
    "            df_gene_reaction_rule[\"Array_67_80°C\"].loc[rxn] = 1000\n",
    "                             \n",
    "    elif rxn == \"oxp_redox_1.6.5.3_RXN0__5330\":\n",
    "        g = df_gene_reaction_rule[\"genes\"].loc[rxn].split(\"or\")\n",
    "        gene_rules = []\n",
    "        gene_numbers1 = []\n",
    "        gene_numbers2 = []\n",
    "        c1 = g[0].split(\",\")\n",
    "        for i in range(len(c1)):\n",
    "        \n",
    "            e = \"SSO\" + c1[i]\n",
    "            if e in df_temperature_treatment.index:\n",
    "                gene_rules.append(e)\n",
    "                gene_numbers1.append(df_temperature_treatment[\"Array_67_70°C\"].loc[e])\n",
    "                gene_numbers2.append(df_temperature_treatment[\"Array_67_80°C\"].loc[e])               \n",
    "            \n",
    "        d1 = str(\" \"+ str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]) + \" \").join(gene_rules)\n",
    "        gene_numbers3 = []\n",
    "        gene_numbers4 = []\n",
    "        c2 = g[1].split(\",\")\n",
    "        for i in range(len(c2)):\n",
    "        \n",
    "            e = \"SSO\" + c2[i]\n",
    "            if e in df_temperature_treatment.index:\n",
    "                gene_rules.append(e)\n",
    "                gene_numbers3.append(df_temperature_treatment[\"Array_67_70°C\"].loc[e])\n",
    "                gene_numbers4.append(df_temperature_treatment[\"Array_67_80°C\"].loc[e])\n",
    "\n",
    "        df_gene_reaction_rule[\"Array_67_70°C\"].loc[rxn] = min(gene_numbers1) + min(gene_numbers3)\n",
    "\n",
    "        df_gene_reaction_rule[\"Array_67_80°C\"].loc[rxn] = min(gene_numbers2) + min(gene_numbers4)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### excel that contains the temperature treatment  transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dynamic_shift =pd.read_excel('../data/transciptone.xlsx',sheet_name = \"dynamic shift\")\n",
    "df_dynamic_shift = df_dynamic_shift.set_index ([\"ORF\"])\n",
    "# df_gene_reaction_rule[\"SSO_5 vs. -10\"] = \"\"\n",
    "df_gene_reaction_rule[\"RAS_5 vs. -10\"] = \"\"\n",
    "# df_gene_reaction_rule[\"SSO_30 vs. 5\"] = \"\"\n",
    "df_gene_reaction_rule[\"RAS_30 vs. 5\"] = \"\"\n",
    "# df_gene_reaction_rule[\"SSO_60 vs. 30\"] = \"\"\n",
    "df_gene_reaction_rule[\"RAS_60 vs. 30\"] = \"\"\n",
    "# df_gene_reaction_rule[\"SSO_60 vs.  -10\"] = \"\"\n",
    "df_gene_reaction_rule[\"RAS_60 vs.  -10\"] = \"\"\n",
    "df_gene_reaction_rule[\"gene_reaction_relation\"] = \"\"\n",
    "for rxn in df_gene_reaction_rule.index:\n",
    "    \n",
    "    if rxn != \"oxp_redox_1.6.5.3_RXN0__5330\":\n",
    "        if 'spontaneous' not in str(df_gene_reaction_rule['genes'].loc[rxn]):\n",
    "            c = df_gene_reaction_rule[\"genes\"].loc[rxn].split(\",\")\n",
    "            if len(c) > 1:\n",
    "                gene_rules = []\n",
    "                gene_numbers1 = []\n",
    "                gene_numbers2 = []\n",
    "                gene_numbers3 = []\n",
    "                gene_numbers4 = []\n",
    "                for i in range(len(c)):\n",
    "                    e = \"SSO\" + c[i]\n",
    "                    if e in df_dynamic_shift.index: \n",
    "                        gene_rules.append(e)\n",
    "                       \n",
    "                        gene_numbers1.append(df_dynamic_shift[\"5 vs.  -10\"].loc[e])\n",
    "                        gene_numbers2.append(df_dynamic_shift[\"30 vs. 5 \"].loc[e])\n",
    "                        gene_numbers3.append(df_dynamic_shift[\"60 vs. 30\"].loc[e])\n",
    "                        gene_numbers4.append(df_dynamic_shift[\"60 vs.  -10\"].loc[e])\n",
    "                if \"or\" in df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]:\n",
    "                    df_gene_reaction_rule[\"RAS_5 vs. -10\"].loc[rxn] = sum(gene_numbers1)\n",
    "                    df_gene_reaction_rule[\"RAS_30 vs. 5\"].loc[rxn] = sum(gene_numbers2)\n",
    "                    df_gene_reaction_rule[\"RAS_60 vs. 30\"].loc[rxn] = sum(gene_numbers3)\n",
    "                    df_gene_reaction_rule[\"RAS_60 vs.  -10\"].loc[rxn] = sum(gene_numbers4)\n",
    "\n",
    "                if \"and\" in str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]):\n",
    "                    df_gene_reaction_rule[\"RAS_5 vs. -10\"].loc[rxn] = min(gene_numbers1)\n",
    "                    df_gene_reaction_rule[\"RAS_30 vs. 5\"].loc[rxn] = min(gene_numbers2)\n",
    "                    df_gene_reaction_rule[\"RAS_60 vs. 30\"].loc[rxn] = min(gene_numbers3)\n",
    "                    df_gene_reaction_rule[\"RAS_60 vs.  -10\"].loc[rxn] = min(gene_numbers4)\n",
    "                d = str(\" \"+ str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]) + \" \").join(gene_rules)\n",
    "                \n",
    "                df_gene_reaction_rule[\"gene_reaction_relation\"].loc[rxn] = \"(\" + d + ')'\n",
    "            elif len(c) == 1:   \n",
    "                e = \"SSO\" + c[0]\n",
    "                if e in df_dynamic_shift.index:\n",
    "                    df_gene_reaction_rule[\"gene_reaction_relation\"].loc[rxn] = e\n",
    "                    df_gene_reaction_rule[\"RAS_5 vs. -10\"].loc[rxn] = df_dynamic_shift[\"5 vs.  -10\"].loc[e]\n",
    "                    df_gene_reaction_rule[\"RAS_30 vs. 5\"].loc[rxn] = df_dynamic_shift[\"30 vs. 5 \"].loc[e]\n",
    "                    df_gene_reaction_rule[\"RAS_60 vs. 30\"].loc[rxn] = df_dynamic_shift[\"60 vs. 30\"].loc[e]\n",
    "                    df_gene_reaction_rule[\"RAS_60 vs.  -10\"].loc[rxn] = df_dynamic_shift[\"60 vs.  -10\"].loc[e]\n",
    "                    \n",
    "        else:\n",
    "            \n",
    "            df_gene_reaction_rule[\"RAS_5 vs. -10\"].loc[rxn] = 1000    \n",
    "            df_gene_reaction_rule[\"RAS_30 vs. 5\"].loc[rxn] = 1000 \n",
    "            df_gene_reaction_rule[\"RAS_60 vs. 30\"].loc[rxn] = 1000 \n",
    "            df_gene_reaction_rule[\"RAS_60 vs.  -10\"].loc[rxn] = 1000   \n",
    "    elif rxn == \"oxp_redox_1.6.5.3_RXN0__5330\":\n",
    "        g = df_gene_reaction_rule[\"genes\"].loc[rxn].split(\"or\")\n",
    "        gene_rules = []\n",
    "        gene_numbers1 = []\n",
    "        gene_numbers2 = []\n",
    "        gene_numbers3 = []\n",
    "        gene_numbers4 = []\n",
    "        c1 = g[0].split(\",\")\n",
    "        for i in range(len(c1)):\n",
    "        \n",
    "            e = \"SSO\" + c1[i]\n",
    "            if e in df_dynamic_shift.index:  \n",
    "                gene_rules.append(e)\n",
    "                gene_numbers1.append(df_dynamic_shift[\"5 vs.  -10\"].loc[e])\n",
    "                gene_numbers2.append(df_dynamic_shift[\"30 vs. 5 \"].loc[e])\n",
    "                gene_numbers3.append(df_dynamic_shift[\"60 vs. 30\"].loc[e])\n",
    "                gene_numbers4.append(df_dynamic_shift[\"60 vs.  -10\"].loc[e])\n",
    "                \n",
    "        d1 = str(\" \"+ str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]) + \" \").join(gene_rules)\n",
    "        \n",
    "        c2 = g[1].split(\",\")\n",
    "        for i in range(len(c2)):\n",
    "            gene_numbers5 = []\n",
    "            gene_numbers6 = []\n",
    "            gene_numbers7 = []\n",
    "            gene_numbers8 = []\n",
    "            e = \"SSO\" + c2[i]\n",
    "            if e in df_dynamic_shift.index:  \n",
    "                gene_numbers5.append(df_dynamic_shift[\"5 vs.  -10\"].loc[e])\n",
    "                gene_numbers6.append(df_dynamic_shift[\"30 vs. 5 \"].loc[e])\n",
    "                gene_numbers7.append(df_dynamic_shift[\"60 vs. 30\"].loc[e])\n",
    "                gene_numbers8.append(df_dynamic_shift[\"60 vs.  -10\"].loc[e])\n",
    "\n",
    "               \n",
    "        \n",
    "        d2 = str(\" \"+ str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]) + \" \").join(gene_rules)\n",
    "        \n",
    "        df_gene_reaction_rule[\"gene_reaction_relation\"].loc[rxn] = \"((\" + d1 + \") or (\" +  d2+ \"))\" \n",
    "        df_gene_reaction_rule[\"RAS_5 vs. -10\"].loc[rxn] = min(gene_numbers1) + min(gene_numbers5)\n",
    "        df_gene_reaction_rule[\"RAS_30 vs. 5\"].loc[rxn] = min(gene_numbers2) + min(gene_numbers6)\n",
    "        df_gene_reaction_rule[\"RAS_60 vs. 30\"].loc[rxn] = min(gene_numbers3) + min(gene_numbers7)\n",
    "        df_gene_reaction_rule[\"RAS_60 vs.  -10\"].loc[rxn] = min(gene_numbers4) + min(gene_numbers8)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RAS_5 vs.  -10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model1= M.copy()\n",
    "df = pd.DataFrame()\n",
    "df[\"soxM\"] = \"\"\n",
    "df[\"soxABCD\"] = \"\"\n",
    "df[\"doxABCDE\"] = \"\"\n",
    "df[\"soxM percent\"] = \"\"\n",
    "df[\"soxABCD percent\"] = \"\"\n",
    "df[\"doxABCDE percent\"] = \"\"\n",
    "for a in [1,3,5,7,9,11,13,15,100]:\n",
    "    for rxn in df_gene_reaction_rule.index: \n",
    "        if \".\" in str(df_gene_reaction_rule['RAS_5 vs. -10'].loc[rxn]):\n",
    "\n",
    "            the_bound = a*df_gene_reaction_rule['RAS_5 vs. -10'].loc[rxn]\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, the_bound)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, the_bound)  # the lower bound will stay zero\n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, 0) # the upper bound will stay zero\n",
    "        else:\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 1000)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            if model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, 1000)  # the lower bound will stay zero\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 0) \n",
    "    model1.objective  = model1.reactions.NGAM\n",
    "    sol = model1.optimize()\n",
    "    df.loc[a] = [sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'])]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RAS_30 vs. 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model1= M.copy()\n",
    "df = pd.DataFrame()\n",
    "df[\"soxM\"] = \"\"\n",
    "df[\"soxABCD\"] = \"\"\n",
    "df[\"doxABCDE\"] = \"\"\n",
    "df[\"soxM percent\"] = \"\"\n",
    "df[\"soxABCD percent\"] = \"\"\n",
    "df[\"doxABCDE percent\"] = \"\"\n",
    "for a in [1,3,5,7,9,11,13,15,100]:\n",
    "    for rxn in df_gene_reaction_rule.index: \n",
    "        if \".\" in str(df_gene_reaction_rule['RAS_30 vs. 5'].loc[rxn]):\n",
    "\n",
    "            the_bound = a*df_gene_reaction_rule['RAS_30 vs. 5'].loc[rxn]\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, the_bound)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, the_bound)  # the lower bound will stay zero\n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, 0) # the upper bound will stay zero\n",
    "        else:\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 1000)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            if model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, 1000)  # the lower bound will stay zero\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 0) \n",
    "    model1.objective  = model1.reactions.NGAM\n",
    "    sol = model1.optimize()\n",
    "    df.loc[a] = [sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'])]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RAS_60 vs. 30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model1= M.copy()\n",
    "df = pd.DataFrame()\n",
    "df[\"soxM\"] = \"\"\n",
    "df[\"soxABCD\"] = \"\"\n",
    "df[\"doxABCDE\"] = \"\"\n",
    "df[\"soxM percent\"] = \"\"\n",
    "df[\"soxABCD percent\"] = \"\"\n",
    "df[\"doxABCDE percent\"] = \"\"\n",
    "for a in [1,3,5,7,9,11,13,15,100]:\n",
    "    for rxn in df_gene_reaction_rule.index: \n",
    "        if \".\" in str(df_gene_reaction_rule['RAS_60 vs. 30'].loc[rxn]):\n",
    "\n",
    "            the_bound = a*df_gene_reaction_rule['RAS_60 vs. 30'].loc[rxn]\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, the_bound)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, the_bound)  # the lower bound will stay zero\n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, 0) # the upper bound will stay zero\n",
    "        else:\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 1000)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            if model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, 1000)  # the lower bound will stay zero\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 0) \n",
    "    model1.objective  = model1.reactions.NGAM\n",
    "    sol = model1.optimize()\n",
    "    df.loc[a] = [sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'])]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RAS_60 vs.  -10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model1= M.copy()\n",
    "df = pd.DataFrame()\n",
    "df[\"soxM\"] = \"\"\n",
    "df[\"soxABCD\"] = \"\"\n",
    "df[\"doxABCDE\"] = \"\"\n",
    "df[\"soxM percent\"] = \"\"\n",
    "df[\"soxABCD percent\"] = \"\"\n",
    "df[\"doxABCDE percent\"] = \"\"\n",
    "for a in [1,3,5,7,9,11,13,15,100]:\n",
    "    for rxn in df_gene_reaction_rule.index: \n",
    "        if \".\" in str(df_gene_reaction_rule['RAS_60 vs.  -10'].loc[rxn]):\n",
    "\n",
    "            the_bound = a*df_gene_reaction_rule['RAS_60 vs.  -10'].loc[rxn]\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, the_bound)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, the_bound)  # the lower bound will stay zero\n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, 0) # the upper bound will stay zero\n",
    "        else:\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 1000)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            if model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, 1000)  # the lower bound will stay zero\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 0) \n",
    "    model1.objective  = model1.reactions.NGAM\n",
    "    sol = model1.optimize()\n",
    "    df.loc[a] = [sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'])]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Different oxygen treatment transcript\n",
    "- The data is from the paper 'Effect of O2 concentrations on Sulfolobus solfataricus P2'.\n",
    "- Sulfolobus solfataricus P2 was grown aerobically, with O2 concentrations ranging from 1.5 to 26 % (v/v; gas phase). To gain some insight in control of the respiratory system, transcriptomes of the strain cultivated in different O2 concentrations (1.5 % vs 21 %, 1.5 % vs 26 %) were compared using a DNA microarray approach.\n",
    "- The data can be download from the website (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE16043)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### excel that contains the oxygen treatment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import the excel file that contain the oxygen treatment transcript data.\n",
    "df_oxygen_transcript =pd.read_excel('../data/transciptone.xlsx',sheet_name = \"Oxygen treatment\")\n",
    "df_oxygen_transcript = df_oxygen_transcript.loc[df_oxygen_transcript['gene'].str.contains('SSO') == True] # Drop out the rows without SSO numbers.\n",
    "df_oxygen_transcript[\"ORF\"] = \"\"\n",
    "df_oxygen_transcript[\"position\"] = \"\"\n",
    "for i in range(len(df_oxygen_transcript)):\n",
    "    df_oxygen_transcript['ORF'].iat[i] = df_oxygen_transcript[\"gene\"].iloc[i].split(\"_\")[0]\n",
    "    df_oxygen_transcript[\"position\"].iat[i] = df_oxygen_transcript[\"gene\"].iloc[i].split(\"_\")[1].split(\" \")[0]\n",
    "df_oxygen_transcript = df_oxygen_transcript.set_index([\"ORF\"],)\n",
    "df_oxygen_transcript[\"1.5vs21\"] = df_oxygen_transcript[[\"1.5vs21 hyb_1\",\"1.5vs21 hyb_2\",\"1.5vs21 hyb_3\",\"21vs1.5 hyb_1\",'21vs1.5 hyb_2',\"21vs1.5 hyb_3\"]].mean(axis = 1)\n",
    "df_oxygen_transcript[\"1.5vs26\"] = df_oxygen_transcript[[\"1.5vs26 hyb_1\",\"1.5vs26 hyb_2\",\"1.5vs26 hyb_3\",\"26vs1.5 hyb_1\",'26vs1.5 hyb_2',\"26vs1.5 hyb_3\"]].mean(axis = 1)\n",
    "\n",
    "df_gene_reaction_rule[\"Oxygen_treatment_1.5vs26 hyb transcript\"] = \"\"\n",
    "df_gene_reaction_rule[\"Oxygen_treatment_1.5vs21 hyb transcript\"] = \"\"\n",
    "df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs26 hyb transcript\"] = \"\"\n",
    "df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs21 hyb transcript\"] = \"\"\n",
    "for rxn in df_gene_reaction_rule.index:  \n",
    "    if rxn != \"oxp_redox_1.6.5.3_RXN0__5330\":\n",
    "        if 'spontaneous' not in str(df_gene_reaction_rule['genes'].loc[rxn]):\n",
    "            c = df_gene_reaction_rule[\"genes\"].loc[rxn].split(\",\")\n",
    "            if len(c) > 1:\n",
    "                gene_numbers1 = []\n",
    "                gene_numbers2 = []\n",
    "                \n",
    "                for i in range(len(c)):\n",
    "                    e = \"SSO\" + c[i]\n",
    "                    if e in df_oxygen_transcript.index:\n",
    "#                        \n",
    "                        gene_numbers1.append(df_oxygen_transcript[\"1.5vs26\"].loc[e])\n",
    "                        gene_numbers2.append(df_oxygen_transcript[\"1.5vs21\"].loc[e])    \n",
    "                        \n",
    "                df_gene_reaction_rule[\"Oxygen_treatment_1.5vs26 hyb transcript\"].loc[rxn] = gene_numbers1\n",
    "                df_gene_reaction_rule[\"Oxygen_treatment_1.5vs21 hyb transcript\"].loc[rxn] = gene_numbers2\n",
    "                \n",
    "                if \"or\" in df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]:\n",
    "                    df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs26 hyb transcript\"].loc[rxn] = sum(gene_numbers1)\n",
    "                    df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs21 hyb transcript\"].loc[rxn] = sum(gene_numbers2)\n",
    "\n",
    "                if \"and\" in str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]):\n",
    "                    df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs26 hyb transcript\"].loc[rxn] = min(gene_numbers1)\n",
    "                    df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs21 hyb transcript\"].loc[rxn] = min(gene_numbers2)\n",
    "                \n",
    "            elif len(c) == 1:\n",
    "                gene_numbers1 = []\n",
    "                gene_numbers2 = []\n",
    "                \n",
    "                e = \"SSO\" + c[0]\n",
    "                if e in df_oxygen_transcript.index:\n",
    "                    gene_numbers1.append(df_oxygen_transcript[\"1.5vs26\"].loc[e])\n",
    "                    gene_numbers2.append(df_oxygen_transcript[\"1.5vs21\"].loc[e])    \n",
    "                        \n",
    "                df_gene_reaction_rule[\"Oxygen_treatment_1.5vs26 hyb transcript\"].loc[rxn] = gene_numbers1\n",
    "                df_gene_reaction_rule[\"Oxygen_treatment_1.5vs21 hyb transcript\"].loc[rxn] = gene_numbers2\n",
    "                df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs26 hyb transcript\"].loc[rxn] = min(gene_numbers1)\n",
    "                df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs21 hyb transcript\"].loc[rxn] = min(gene_numbers2)\n",
    "        else:\n",
    "            \n",
    "            df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs26 hyb transcript\"].loc[rxn] = 1000\n",
    "            df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs21 hyb transcript\"].loc[rxn] = 1000        \n",
    "    elif rxn == \"oxp_redox_1.6.5.3_RXN0__5330\":\n",
    "        g = df_gene_reaction_rule[\"genes\"].loc[rxn].split(\"or\")\n",
    "        gene_numbers1 = []\n",
    "        gene_numbers2 = []\n",
    "        \n",
    "        c1 = g[0].split(\",\")\n",
    "        for i in range(len(c1)):\n",
    "        \n",
    "            e = \"SSO\" + c1[i]\n",
    "            if e in df_oxygen_transcript.index:\n",
    "\n",
    "                gene_numbers1.append(df_oxygen_transcript[\"1.5vs26\"].loc[e])\n",
    "                gene_numbers2.append(df_oxygen_transcript[\"1.5vs21\"].loc[e]) \n",
    "#             if \"or\" in df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]:\n",
    "#                 df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs26 hyb transcript\"].loc[rxn] = sum(gene_numbers1)\n",
    "#                 df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs21 hyb transcript\"].loc[rxn] = sum(gene_numbers2)\n",
    "\n",
    "#             if \"and\" in str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]):\n",
    "#                 df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs26 hyb transcript\"].loc[rxn] = min(gene_numbers1)\n",
    "#                 df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs21 hyb transcript\"].loc[rxn] = min(gene_numbers2)\n",
    "        gene_numbers3 = []\n",
    "        gene_numbers4 = []        \n",
    "        c2 = g[1].split(\",\")\n",
    "        for i in range(len(c2)):\n",
    "            e = \"SSO\" + c2[i]\n",
    "            if e in df_oxygen_transcript.index:\n",
    "                gene_numbers3.append(df_oxygen_transcript[\"1.5vs26\"].loc[e])\n",
    "                gene_numbers4.append(df_oxygen_transcript[\"1.5vs21\"].loc[e]) \n",
    "\n",
    "        df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs26 hyb transcript\"].loc[rxn] = min(gene_numbers1) + min(gene_numbers2)\n",
    "        df_gene_reaction_rule[\"RAS_Oxygen_treatment_1.5vs21 hyb transcript\"].loc[rxn] = min(gene_numbers2) + min(gene_numbers4)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Oxygen_treatment_1.5vs21 hyb transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model1= M.copy()\n",
    "df = pd.DataFrame()\n",
    "df[\"soxM\"] = \"\"\n",
    "df[\"soxABCD\"] = \"\"\n",
    "df[\"doxABCDE\"] = \"\"\n",
    "df[\"soxM percent\"] = \"\"\n",
    "df[\"soxABCD percent\"] = \"\"\n",
    "df[\"doxABCDE percent\"] = \"\"\n",
    "for a in [1,3,5,7,9,11,13,15,100,1000]:\n",
    "    for rxn in df_gene_reaction_rule.index: \n",
    "        if \".\" in str(df_gene_reaction_rule['RAS_Oxygen_treatment_1.5vs21 hyb transcript'].loc[rxn]):\n",
    "\n",
    "            the_bound = a*df_gene_reaction_rule['RAS_Oxygen_treatment_1.5vs21 hyb transcript'].loc[rxn]\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, the_bound)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, the_bound)  # the lower bound will stay zero\n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, 0) # the upper bound will stay zero\n",
    "        else:\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 1000)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            if model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, 1000)  # the lower bound will stay zero\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 0) \n",
    "    model1.objective  = model1.reactions.NGAM\n",
    "    sol = model1.optimize()\n",
    "    df.loc[a] = [sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'])]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for rxn in df_gene_reaction_rule.index: \n",
    "    if \".\" in str(df_gene_reaction_rule['RAS_Oxygen_treatment_1.5vs26 hyb transcript'].loc[rxn]):\n",
    "\n",
    "        the_bound = 100*df_gene_reaction_rule['RAS_Oxygen_treatment_1.5vs26 hyb transcript'].loc[rxn]\n",
    "\n",
    "        if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, the_bound)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "        elif model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (0, the_bound)  # the lower bound will stay zero\n",
    "\n",
    "        elif model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, 0) # the upper bound will stay zero\n",
    "    else:\n",
    "        if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (-1000, 1000)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "        if model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (0, 1000)  # the lower bound will stay zero\n",
    "\n",
    "        if model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (-1000, 0) \n",
    "model1.objective  = model1.reactions.NGAM\n",
    "sol = model1.optimize()\n",
    "print (\"NGAM\", sol.fluxes.loc['NGAM'])\n",
    "b = show_map(sol,map_loc)\n",
    "b"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Oxygen_treatment_1.5vs26 hyb transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model1= M.copy()\n",
    "df = pd.DataFrame()\n",
    "df[\"soxM\"] = \"\"\n",
    "df[\"soxABCD\"] = \"\"\n",
    "df[\"doxABCDE\"] = \"\"\n",
    "df[\"soxM percent\"] = \"\"\n",
    "df[\"soxABCD percent\"] = \"\"\n",
    "df[\"doxABCDE percent\"] = \"\"\n",
    "for a in [1,3,5,7,9,11,13,15,100,1000]:\n",
    "    for rxn in df_gene_reaction_rule.index: \n",
    "        if \".\" in str(df_gene_reaction_rule['RAS_Oxygen_treatment_1.5vs26 hyb transcript'].loc[rxn]):\n",
    "\n",
    "            the_bound = a*df_gene_reaction_rule['RAS_Oxygen_treatment_1.5vs26 hyb transcript'].loc[rxn]\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, the_bound)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, the_bound)  # the lower bound will stay zero\n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, 0) # the upper bound will stay zero\n",
    "        else:\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 1000)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            if model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, 1000)  # the lower bound will stay zero\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 0) \n",
    "    model1.objective  = model1.reactions.NGAM\n",
    "    sol = model1.optimize()\n",
    "    df.loc[a] = [sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'])]    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add tge gene reaction rule to the model\n",
    "# reactions_ids = [rxn.id for rxn in model.reactions]\n",
    "df_gene_reaction_rule[\"SSO_5 vs. -10\"] = \"\"\n",
    "df_gene_reaction_rule[\"RAS_5 vs. -10\"] = \"\"\n",
    "\n",
    "df_gene_reaction_rule[\"gene_reaction_relation\"] = \"\"\n",
    "for rxn in df_gene_reaction_rule.index:\n",
    "    \n",
    "    if rxn != \"oxp_redox_1.6.5.3_RXN0__5330\":\n",
    "        if 'spontaneous' not in str(df_gene_reaction_rule['genes'].loc[rxn]):\n",
    "            c = df_gene_reaction_rule[\"genes\"].loc[rxn].split(\",\")\n",
    "            if len(c) > 1:\n",
    "                gene_rules = []\n",
    "                gene_numbers1 = []\n",
    "                gene_numbers2 = []\n",
    "                \n",
    "                for i in range(len(c)):\n",
    "                    e = \"SSO\" + c[i]\n",
    "                    if e in df_dynamic_shift.index: \n",
    "                        gene_rules.append(e)\n",
    "#                         ge = df_gene_reaction_rule[\"SSO_5 vs. -10\"].loc[e]\n",
    "                       \n",
    "                        gene_numbers1.append(df_dynamic_shift[\"5 vs.  -10\"].loc[e])\n",
    "                if \"or\" in df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]:\n",
    "                    df_gene_reaction_rule[\"RAS_5 vs. -10\"].loc[rxn] = sum(gene_numbers1)\n",
    "\n",
    "#                     df_gene_reaction_rule[\"RAS_control\"].loc[rxn] = sum(gene_numbers2)\n",
    "                if \"and\" in str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]):\n",
    "                    df_gene_reaction_rule[\"RAS_5 vs. -10\"].loc[rxn] = min(gene_numbers1)\n",
    "\n",
    "#                     df_gene_reaction_rule[\"RAS_control\"].loc[rxn] = min(gene_numbers2)\n",
    "                    \n",
    "\n",
    "        \n",
    "                d = str(\" \"+ str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]) + \" \").join(gene_rules)\n",
    "                df_gene_reaction_rule[\"SSO_5 vs. -10\"].loc[rxn] = gene_numbers1\n",
    "#                 df_gene_reaction_rule[\"control\"].loc[rxn] = gene_numbers2\n",
    "                df_gene_reaction_rule[\"gene_reaction_relation\"].loc[rxn] = \"(\" + d + ')'\n",
    "            elif len(c) == 1:\n",
    "                gene_rules = []\n",
    "                gene_numbers1 = []\n",
    "                \n",
    "                e = \"SSO\" + c[0]\n",
    "                if e in df_dynamic_shift.index:\n",
    "                    d = e \n",
    "                    gene_numbers1.append(df_dynamic_shift[\"5 vs.  -10\"].loc[e])                   \n",
    "                    df_gene_reaction_rule[\"SSO_5 vs. -10\"].loc[rxn] = gene_numbers1\n",
    "                    df_gene_reaction_rule[\"gene_reaction_relation\"].loc[rxn] = d\n",
    "                    df_gene_reaction_rule[\"RAS_5 vs. -10\"].loc[rxn] = df_dynamic_shift[\"5 vs.  -10\"].loc[e]\n",
    "        else:\n",
    "            \n",
    "            df_gene_reaction_rule[\"SSO_5 vs. -10\"].loc[rxn] = 1000\n",
    "                    \n",
    "           \n",
    "    elif rxn == \"oxp_redox_1.6.5.3_RXN0__5330\":\n",
    "        g = df_gene_reaction_rule[\"genes\"].loc[rxn].split(\"or\")\n",
    "        gene_rules = []\n",
    "        gene_numbers1 = []\n",
    "        gene_numbers2 = []\n",
    "        c1 = g[0].split(\",\")\n",
    "        for i in range(len(c1)):\n",
    "        \n",
    "            e = \"SSO\" + c1[i]\n",
    "            if e in df_dynamic_shift.index:  \n",
    "                gene_rules.append(e)\n",
    "                gene_numbers1.append(df_dynamic_shift[\"5 vs.  -10\"].loc[e])\n",
    "\n",
    "        \n",
    "               \n",
    "        \n",
    "        d1 = str(\" \"+ str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]) + \" \").join(gene_rules)\n",
    "        gene_numbers3 = []\n",
    "        gene_numbers4 = []\n",
    "        c2 = g[1].split(\",\")\n",
    "        for i in range(len(c2)):\n",
    "        \n",
    "            e = \"SSO\" + c2[i]\n",
    "            if e in df_dynamic_shift.index:  \n",
    "                gene_numbers3.append(df_dynamic_shift[\"5 vs.  -10\"].loc[e])\n",
    "\n",
    "               \n",
    "        df_gene_reaction_rule[\"SSO_5 vs. -10\"].loc[rxn] = gene_numbers1 + gene_numbers3\n",
    "\n",
    "        d2 = str(\" \"+ str(df_gene_reaction_rule['gene_reaction_rule'].loc[rxn]) + \" \").join(gene_rules)\n",
    "        \n",
    "        df_gene_reaction_rule[\"gene_reaction_relation\"].loc[rxn] = \"((\" + d1 + \") or (\" +  d2+ \"))\" \n",
    "        df_gene_reaction_rule[\"RAS_5 vs. -10\"].loc[rxn] = min(gene_numbers1) + min(gene_numbers3)\n",
    "#         df_gene_reaction_rule[\"RAS_control\"].loc[rxn] = min(gene_numbers2) + min(gene_numbers4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene_reaction_rule.to_csv('../data/oxygen_treatment2020.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The gearshifting among different electron transfer chain "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model1 = model.copy()\n",
    "model1.reactions.NGAM.bounds = (0,1000)\n",
    "model1.objective = model1.reactions.NGAM\n",
    "sol = model1.optimize()\n",
    "l = sol.fluxes[abs(sol.fluxes)>1e-6].index\n",
    "# for rxn in l:\n",
    "#     print (rxn, sol.fluxes.loc[rxn])\n",
    "\n",
    "for r in model1.reactions:\n",
    "    if r.id in l:\n",
    "#         print (r.id)\n",
    "        if sol.fluxes.loc[r.id] > 0:\n",
    "            model1.reactions.get_by_id(r.id).upper_bound = sol.fluxes.loc[r.id]\n",
    "           \n",
    "        elif sol.fluxes.loc[r.id] < 0:\n",
    "            model1.reactions.get_by_id(r.id).lower_bound = sol.fluxes.loc[r.id]\n",
    "          \n",
    "    else:\n",
    "        model1.reactions.get_by_id(r.id).bounds = (0,0)\n",
    "b = show_map(sol,map_loc)\n",
    "b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame()\n",
    "df[\"soxM\"] = \"\"\n",
    "df[\"soxABCD\"] = \"\"\n",
    "df[\"doxABCDE\"] = \"\"\n",
    "df[\"soxM percent\"] = \"\"\n",
    "df[\"soxABCD percent\"] = \"\"\n",
    "df[\"doxABCDE percent\"] = \"\"\n",
    "for a in [1,3,5,7,9,11,13,15]:\n",
    "    for rxn in df_gene_reaction_rule.index: \n",
    "        if \".\" in str(df_gene_reaction_rule['RAS_HU TREATED/CONTROL'].loc[rxn]):\n",
    "\n",
    "            the_bound = a*df_gene_reaction_rule['RAS_HU TREATED/CONTROL'].loc[rxn]\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, the_bound)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, the_bound)  # the lower bound will stay zero\n",
    "\n",
    "            elif model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, 0) # the upper bound will stay zero\n",
    "        else:\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 1000)# we set the bound(s) with max abs(FVA) \n",
    "\n",
    "            if model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (0, 1000)  # the lower bound will stay zero\n",
    "\n",
    "            if model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "\n",
    "                model1.reactions.get_by_id(rxn).bounds = (-1000, 0) \n",
    "    model1.objective  = model1.reactions.NGAM\n",
    "    sol = model1.optimize()\n",
    "    df.loc[a] = [sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'],sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240'] + sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']),sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240']/(sol.fluxes.loc['oxp_redox_1.10.3.13_soxM_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_soxABCD_TRANS__RXN__240']+sol.fluxes.loc['oxp_redox_1.10.3.13_doxABCDE_TRANS__RXN__240'])]    \n",
    "    \n",
    "#     l = sol.fluxes[abs(sol.fluxes)>1e-6].index\n",
    "#     for rxn in l:\n",
    "#         print (rxn, sol.fluxes.loc[rxn])\n",
    "#     # print (\"NGAM\", sol.fluxes.loc['NGAM'])\n",
    "#     b = show_map(sol,map_loc)\n",
    "#     b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model2 = model.copy()\n",
    "for rxn in df_gene_reaction_rule.index:\n",
    "    the_bound = df_gene_reaction_rule['RAS_control'].loc[rxn]/10\n",
    "\n",
    "    if model.reactions.get_by_id(rxn).lower_bound < 0 and model.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "       \n",
    "        model2.reactions.get_by_id(rxn).bounds = (-1*the_bound , the_bound) # we set the bound(s) with max abs(FVA) \n",
    "     \n",
    "    elif model.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "        \n",
    "        model2.reactions.get_by_id(rxn).bounds = (0, the_bound)  # the lower bound will stay zero\n",
    "      \n",
    "    elif model.reactions.get_by_id(rxn).lower_bound < 0: # irreversible, flows in backward direction\n",
    "        \n",
    "        model2.reactions.get_by_id(rxn).bounds = (-1*the_bound, 0) # the upper bound will stay zero\n",
    "        \n",
    "      \n",
    "\n",
    "    model2.repair()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M = model2.copy()\n",
    "M.objective  = M.reactions.NGAM\n",
    "for rxn in M.reactions:\n",
    "    if \"T_flux_L\" in rxn.id:\n",
    "        rxn.bounds = (0,0)\n",
    "sol = M.optimize()\n",
    "l = sol.fluxes[abs(sol.fluxes)>1e-6].index\n",
    "for rxn in l:   \n",
    "    print (rxn,'\\t',sol.fluxes.loc[rxn],'\\t',model.reactions.get_by_id(rxn).reaction, '\\n')\n",
    "b = show_map(sol,map_loc)\n",
    "b\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model1 = M.copy()\n",
    "\n",
    "for rxn in df_gene_reaction_rule.index:\n",
    "    df = pd.DataFrame(columns=[\"NGAM\"])\n",
    "    list = [1, 10, 100, 1000,10000]\n",
    "    for a in list:\n",
    "        \n",
    "        the_bound = df_gene_reaction_rule['RAS_HU'].loc[rxn]/a\n",
    "\n",
    "        if model1.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, the_bound) # we set the bound(s) with max abs(FVA) \n",
    "\n",
    "        elif model1.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (0, the_bound)  # the lower bound will stay zero\n",
    "\n",
    "        elif int(model1.reactions.get_by_id(rxn).lower_bound) < 0: # irreversible, flows in backward direction\n",
    "\n",
    "            model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, 0) # the upper bound will stay zero\n",
    "        model1.objective  = model1.reactions.NGAM\n",
    "        sol = model1.optimize()\n",
    "    df.loc[a] = [-sol.fluxes.loc[\"NGAM\"]] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for rxn in df_gene_reaction_rule.index:        \n",
    "    the_bound = df_gene_reaction_rule['RAS_control'].loc[rxn]/5\n",
    "\n",
    "    if model1.reactions.get_by_id(rxn).lower_bound < 0 and model1.reactions.get_by_id(rxn).upper_bound > 0: # this finds reversible rxns\n",
    "\n",
    "        model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, the_bound) # we set the bound(s) with max abs(FVA) \n",
    "\n",
    "    elif model1.reactions.get_by_id(rxn).upper_bound > 0: # irreversible, flows in forward direction\n",
    "\n",
    "        model1.reactions.get_by_id(rxn).bounds = (0, the_bound)  # the lower bound will stay zero\n",
    "\n",
    "    elif int(model1.reactions.get_by_id(rxn).lower_bound) < 0: # irreversible, flows in backward direction\n",
    "\n",
    "        model1.reactions.get_by_id(rxn).bounds = (-1*the_bound, 0) # the upper bound will stay zero\n",
    "model1.objective  = model1.reactions.NGAM\n",
    "\n",
    "sol = model1.optimize()\n",
    "print (\"NGAM\",sol.fluxes.loc['NGAM'])\n",
    "b = show_map(sol,map_loc)\n",
    "b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "273.188px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
